File: frontend/src/lib/schemas.ts
import { z } from "zod";

/* ────────────────────────── primitives ────────────────────────── */

export const LanguageSchema = z.enum(["en", "ur", "ar"]);
export type Language = z.infer<typeof LanguageSchema>;

export const ToneKeySchema = z.enum([
  "formal",
  "empathetic",
  "leadership",
  "diplomatic",
  "direct",
  "encouraging",
  "apologetic",
  "persuasive",
]);
export type ToneKey = z.infer<typeof ToneKeySchema>;

/* ─────────────────────────── analyze ──────────────────────────── */

export const AnalyzeRequestSchema = z.object({
  text: z.string().min(1).max(8000),
  language: LanguageSchema,
});
export type AnalyzeRequest = z.infer<typeof AnalyzeRequestSchema>;

export const EmotionSchema = z.object({
  name: z.string(),
  intensity: z.number().min(0).max(100),
});
export type Emotion = z.infer<typeof EmotionSchema>;

export const PersonalitySignalsSchema = z.object({
  assertiveness: z.number().min(0).max(100),
  empathy:       z.number().min(0).max(100),
  clarity:       z.number().min(0).max(100),
  confidence:    z.number().min(0).max(100),
});
export type PersonalitySignals = z.infer<typeof PersonalitySignalsSchema>;

export const AnalyzeResponseSchema = z.object({
  communicationScore: z.number().min(0).max(100),
  currentTone:        z.string(),
  emotions:           z.array(EmotionSchema),
  personality:        PersonalitySignalsSchema,
  strengths:          z.array(z.string()),
  suggestions:        z.array(z.string()),
  refinedMessage:     z.string(),
  refinedTone:        z.string(),
});
export type AnalyzeResponse = z.infer<typeof AnalyzeResponseSchema>;

/* ─────────────────────────── rewrite ──────────────────────────── */

export const RewriteRequestSchema = z.object({
  text:     z.string().min(1).max(8000),
  tone:     ToneKeySchema,
  language: LanguageSchema,
});
export type RewriteRequest = z.infer<typeof RewriteRequestSchema>;

export const RewriteResponseSchema = z.object({
  rewritten: z.string(),
  rationale: z.string(),
});
export type RewriteResponse = z.infer<typeof RewriteResponseSchema>;

/* ───────────────────────── interview ──────────────────────────── */

export const InterviewQuestionRequestSchema = z.object({
  role:     z.string().min(1).max(120),
  language: LanguageSchema,
});
export type InterviewQuestionRequest = z.infer<typeof InterviewQuestionRequestSchema>;

export const InterviewQuestionResponseSchema = z.object({
  question: z.string(),
});
export type InterviewQuestionResponse = z.infer<typeof InterviewQuestionResponseSchema>;

export const InterviewFeedbackRequestSchema = z.object({
  role:     z.string().min(1).max(120),
  question: z.string().min(1).max(2000),
  answer:   z.string().min(1).max(8000),
  language: LanguageSchema,
});
export type InterviewFeedbackRequest = z.infer<typeof InterviewFeedbackRequestSchema>;

export const InterviewFeedbackResponseSchema = z.object({
  score:         z.number().min(0).max(100),
  verdict:       z.string(),
  strengths:     z.array(z.string()),
  improvements:  z.array(z.string()),
  refinedAnswer: z.string(),
});
export type InterviewFeedbackResponse = z.infer<typeof InterviewFeedbackResponseSchema>;

/* ────────────────────────── shared API ────────────────────────── */

export const ErrorResponseSchema = z.object({
  error: z.string(),
  detail: z.string().optional(),
});
export type ErrorResponse = z.infer<typeof ErrorResponseSchema>;

File: backend/app/schemas.py
"""Pydantic v2 schemas — must stay in sync with frontend/src/lib/schemas.ts.

This file is the single source of truth on the backend side. The TypeScript
zod schemas are the mirror used by the Next.js proxy until it is removed.
"""
from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class Language(StrEnum):
    EN = "en"
    UR = "ur"
    AR = "ar"


class ToneKey(StrEnum):
    FORMAL = "formal"
    EMPATHETIC = "empathetic"
    LEADERSHIP = "leadership"
    DIPLOMATIC = "diplomatic"
    DIRECT = "direct"
    ENCOURAGING = "encouraging"
    APOLOGETIC = "apologetic"
    PERSUASIVE = "persuasive"


class _Model(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


# ────────────────────────── analyze ──────────────────────────

class AnalyzeRequest(_Model):
    text: str = Field(min_length=1, max_length=8000)
    language: Language


class Emotion(_Model):
    name: str
    intensity: int = Field(ge=0, le=100)


class PersonalitySignals(_Model):
    assertiveness: int = Field(ge=0, le=100)
    empathy:       int = Field(ge=0, le=100)
    clarity:       int = Field(ge=0, le=100)
    confidence:    int = Field(ge=0, le=100)


class AnalyzeResponse(_Model):
    communicationScore: int = Field(ge=0, le=100)
    currentTone: str
    emotions: list[Emotion]
    personality: PersonalitySignals
    strengths: list[str]
    suggestions: list[str]
    refinedMessage: str
    refinedTone: str


# ────────────────────────── rewrite ──────────────────────────

class RewriteRequest(_Model):
    text: str = Field(min_length=1, max_length=8000)
    tone: ToneKey
    language: Language


class RewriteResponse(_Model):
    rewritten: str
    rationale: str


# ───────────────────────── interview ─────────────────────────

class InterviewQuestionRequest(_Model):
    role: str = Field(min_length=1, max_length=120)
    language: Language


class InterviewQuestionResponse(_Model):
    question: str


class InterviewFeedbackRequest(_Model):
    role: str = Field(min_length=1, max_length=120)
    question: str = Field(min_length=1, max_length=2000)
    answer: str = Field(min_length=1, max_length=8000)
    language: Language


class InterviewFeedbackResponse(_Model):
    score: int = Field(ge=0, le=100)
    verdict: str
    strengths: list[str]
    improvements: list[str]
    refinedAnswer: str


# ──────────────────────── shared API ─────────────────────────

class ErrorResponse(_Model):
    error: str
    detail: str | None = None
